src/middleware/withUcantoClient.js

```diff
@@ -0,0 +1,109 @@
+export function withUcantoClient (handler) {
+  return async (req, env, ctx) => {
+    const ucantoClient = await create(env)
+
+    return handler(req, env, { ...ctx, ucantoClient })
+  }
+}
+
+/**
+ * Creates a UCantoClient instance with the given environment and establishes a connection to the UCanto Server.
+ *
+ * @param {Environment} env
+ * @returns {Promise<import('./withUcantoClient.types.ts').UCantoClient>}
+ */
+async function create (env) {
+  const service = Verifier.parse(env.SERVICE_ID)
+  const principal = Signer.parse(env.SIGNER_PRINCIPAL_KEY)
+  const { connection } = await connect(env.UPLOAD_API_URL, principal)
+
+  return {
+    /**
+     * Records the egress bytes for the given resource.
+     *
+     * @param {import('@ucanto/principal/ed25519').DIDKey} space - The Space DID where the content was served
+     * @param {import('@ucanto/principal/ed25519').UnknownLink} resource - The link to the resource that was served
+     * @param {number} bytes - The number of bytes served
+     * @param {Date} servedAt - The timestamp of when the content was served
+     * @returns {Promise<void>}
+     */
+    record: async (space, resource, bytes, servedAt) =>
+      recordEgress(space, resource, bytes, servedAt, principal, service, connection),
+
+    /**
+     * TODO: implement this function
+     *
+     * @param {string} authToken
+     * @returns {Promise<import('./withUcantoClient.types.ts').TokenMetadata | undefined>}
+     */
+    getTokenMetadata: async (authToken) => {
+      // TODO I think this needs to check the content claims service (?) for any claims relevant to this token
+      // TODO do we have a plan for this? need to ask Hannah if the indexing service covers this?
+      return undefined
+    }
+  }
+}
+
+/**
+ * Creates a connection with the UCanto Server at the provided server URL.
+ *
+ * @param {string} serverUrl
+ * @param {import('@ucanto/principal/ed25519').EdSigner} principal
+ *
+ */
+async function connect (serverUrl, principal) {
+  const connection = await UCantoClient.connect({
+    id: principal,
+    codec: CAR.outbound,
+    channel: HTTP.open({ url: new URL(serverUrl) })
+  })
+
+  return { connection }
+}
+
+/**
+ * Records the egress bytes in the UCanto Server by invoking the `Usage.record` capability.
+ *
+ * @param {import('@ucanto/principal/ed25519').DIDKey} space - The Space DID where the content was served
+ * @param {import('@ucanto/principal/ed25519').UnknownLink} resource - The link to the resource that was served
+ * @param {number} bytes - The number of bytes served
+ * @param {Date} servedAt - The timestamp of when the content was served
+ * @param {import('@ucanto/principal/ed25519').EdSigner} principal - The principal signer
+ * @param {Signer.Verifier} service - The service verifier
+ * @param {any} connection - The connection to execute the command
+ * @returns {Promise<void>}
+ */
+async function recordEgress (space, resource, bytes, servedAt, principal, service, connection) {
```

Comment:
**suggestion (non-blocking):** I think I'd either keep this to being a pure Ucanto client (which I guess is properly termed a Ucanto "connection") and move the invocation into `withRateLimit`, or change the thing in the context from `ucantoClient` to…maybe `usageClient`? A "Ucanto client" wouldn't know about our specific (cap)abilities.

I think the latter makes a lot of sense given the way we've defined "clients" in `w3up`, and it's also the easiest change. 😄

Reply:
Thanks. I've renamed that to `egressClient`.